File: swf-animate/api/FxArtManager.lua

--------------------------------
-- @module FxArtManager
-- @extend FxUnitManager
-- @parent_module flashx

--------------------------------
-- @overload self, flashx.FxAnimateData         
-- @overload self, unsigned int         
-- @function [parent=#FxArtManager] getAnimateRes
-- @param self
-- @param #unsigned int animateId
-- @return array_table#array_table ret (return value: array_table)

--------------------------------
-- 归还资源集<br>
-- param retianRes 引用的动画资源表
-- @function [parent=#FxArtManager] releaseAtlasSet 
-- @param self
-- @param #array_table retianRes
-- @return FxArtManager#FxArtManager self (return value: flashx.FxArtManager)
        
--------------------------------
-- 引用动画资源<br>
-- param animateData 动画数据<br>
-- param retianRes 引用的动画资源表
-- @function [parent=#FxArtManager] retainAnimateRes 
-- @param self
-- @param #flashx.FxAnimateData animateData
-- @param #array_table retianRes
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 初始化<br>
-- param animateDir 动画根目录<br>
-- param actPkgPath 动作数据包路径
-- @function [parent=#FxArtManager] init 
-- @param self
-- @param #string animateDir
-- @param #string actPkgPath
-- @return FxArtManager#FxArtManager self (return value: flashx.FxArtManager)
        
--------------------------------
-- 构建异步动画<br>
-- param animateId 动画ID<br>
-- param handler 动画资源加载回调
-- @function [parent=#FxArtManager] newAsyncAnimate 
-- @param self
-- @param #unsigned int animateId
-- @param #function handler
-- @return FxAsyncAnimate#FxAsyncAnimate ret (return value: flashx.FxAsyncAnimate)
        
--------------------------------
-- 构建动画<br>
-- param animateId 动画ID<br>
-- return 动画
-- @function [parent=#FxArtManager] newAnimate 
-- @param self
-- @param #unsigned int animateId
-- @return FxAnimate#FxAnimate ret (return value: flashx.FxAnimate)
        
--------------------------------
-- 
-- @function [parent=#FxArtManager] disposeInstance 
-- @param self
-- @return FxArtManager#FxArtManager self (return value: flashx.FxArtManager)
        
--------------------------------
-- 
-- @function [parent=#FxArtManager] getInstance 
-- @param self
-- @return FxArtManager#FxArtManager ret (return value: flashx.FxArtManager)
        
return nil

File: swf-animate/art2/manager/FxArtManager.cpp
//
//  FxArtManager.cpp
//
//  Created by joli on 2018/8/22.
//  Copyright ? 2018Äê uzone. All rights reserved.
//

#include "flashx/art2/manager/FxArtManager.h"
#include "flashx/art2/manager/FxDataManager.h"
#include "flashx/manager/FxAtlasManager.h"

NS_FLASHX_BEGIN

static fxstr pathOfPlist(const u32& plistID, const fxstr& dir)
{
    static char buf[64];
    sprintf(buf, "%s/%u.plist", dir.c_str(), plistID);
    return buf;
}

static FxArtManager* singleton = nullptr;
FxArtManager* FxArtManager::getInstance()
{
    if (!singleton) {
        singleton = new (std::nothrow) FxArtManager();
    }
    return singleton;
}

void FxArtManager::disposeInstance()
{
    FX_SAFE_DELETE(singleton);
}

FxArtManager::FxArtManager() {}
FxArtManager::~FxArtManager() {}

void FxArtManager::onExecuteGC() {
#ifdef FX_DEBUG
    FXLOG("************************* START ****************************");
    clock_t startTime = clock();
#endif
    FxAtlasManager::getInstance()->removeUnsuedAtlas();
#ifdef FX_DEBUG
    FXLOG("units remain:%u", numUnits());
    FXLOG("atlas remain:%u", FxAtlasManager::getInstance()->numAtlas());
    FXLOG("gc time:%lf(s)", (double)(clock()-startTime)/CLOCKS_PER_SEC);
    FXLOG("*************************  END  ****************************");
#endif
}

void FxArtManager::init(const fxstr& animateDir, const fxstr& actPkgPath)
{
    animateSheetDir = animateDir + "/sheet";
    animateGroupDir = animateDir + "/group";
    FxDataManager::getInstance()->init(animateDir + "/datas", animateDir + "/SheetPackage.fbf", actPkgPath);
    resumeTimeline();
}

FxAnimate* FxArtManager::newAnimate(const u32& animateId)
{
#ifdef FX_DEBUG
    clock_t startTime = clock();
#endif
    FxAnimate* animate = new FxAnimate();
    if (animate) {
        animate->bindArt(this);
        if (animate->initWithID(animateId)) {
            pushUnit(animate);
            animate->release();
        } else {
            CC_SAFE_RELEASE_NULL(animate);
        }
    }
#ifdef FX_DEBUG
    double endTime = (clock() - startTime) / CLOCKS_PER_SEC;
    FXLOG("%s time:%lf(s)", __func__, endTime);
#endif
    return animate;
}

FxAsyncAnimate* FxArtManager::newAsyncAnimate(const u32& animateId, const FxAnimateLoadHandler& handler /* nullptr */)
{
#ifdef FX_DEBUG
    clock_t startTime = clock();
#endif
    FxAsyncAnimate* animate = new FxAsyncAnimate();
    if (animate) {
        animate->bindArt(this);
        if (animate->initWithLoader(animateId, handler)) {
            pushUnit(animate);
            animate->release();
        } else {
            CC_SAFE_RELEASE_NULL(animate);
        }
    }
#ifdef FX_DEBUG
    FXLOG("%s time:%lf(s)", __func__, (double)(clock()-startTime)/CLOCKS_PER_SEC);
#endif
    return animate;
}

FxStrArray FxArtManager::getAnimateRes(const u32& animateId) {
    FxStrArray resSet;
    const FxAnimateData* animateData = FxDataManager::getInstance()->getAnimateData(animateId);
    if (animateData) {
        resSet = getAnimateRes(animateData);
    }
    return resSet;
}

FxStrArray FxArtManager::getAnimateRes(const FxAnimateData* animateData)
{
    FxStrArray resSet;
    const FxSheetSet* sheetSet = animateData->getAtlasSheets();
    if (sheetSet) {
        const u8& len = sheetSet->getSize();
        for (u8 i=0; i<len; ++i) {
            resSet.push_back(pathOfPlist(sheetSet->getSheetID(i), animateSheetDir));
        }
    }
    sheetSet = animateData->getGroupSheets();
    if (sheetSet) {
        const u8& len = sheetSet->getSize();
        for (u8 i=0; i<len; ++i) {
            resSet.push_back(pathOfPlist(sheetSet->getSheetID(i), animateGroupDir));
        }
    }
    return resSet;
}

bool FxArtManager::retainAnimateRes(const FxAnimateData* animateData, FxStrArray& retianRes)
{
    const std::vector<fxstr>& resSet = getAnimateRes(animateData);
    forit(resSet, resItr) {
        if (!FxAtlasManager::getInstance()->retainAtlas(*resItr)) {
            FXLOG("retian %s failed:%s", __func__, (*resItr).c_str());
            return false;
        }
        retianRes.push_back(*resItr);
    }
    return true;
}
void FxArtManager::releaseAtlasSet(FxStrArray& retianRes)
{
    forit(retianRes, resItr) {
        FxAtlasManager::getInstance()->releaseAtlas(*resItr);
    }
}

NS_FLASHX_END /* NS_FLASHX_BEGIN */
